editoast/src/models/paced_train.rs

```diff
@@ -126,16 +129,75 @@ impl PacedTrain {
         }
     }
 
-    pub fn iter_occurrences(&self) -> impl Iterator<Item = TrainSchedule> {
-        let base_occurrence = self.clone().into_train_schedule();
-
-        (0..self.num_occurrences()).map(move |occurrence_idx| TrainSchedule {
-            start_time: base_occurrence.start_time + self.interval * occurrence_idx as i32,
-            ..base_occurrence.clone()
+    fn find_modified_exception(&self, occurrence_idx: usize) -> Option<&PacedTrainException> {
+        self.exceptions.iter().find(|exception| {
+            matches!(
+                exception.exception_type,
+                ExceptionType::Modified { occurrence_index }
+                    if occurrence_index == occurrence_idx as i32 && !exception.disabled
+            )
         })
     }
 
-    pub fn num_occurrences(&self) -> usize {
+    pub fn iter_occurrences(&self) -> impl Iterator<Item = (TrainId, TrainSchedule)> {
```

Comment:
I find this function quite difficult to read.
Maybe we could add documentation comments. But most important extract some logic of it in other functions.